simulator/main.rs

```diff
@@ -67,6 +123,433 @@ enum Value {
+        }
+
+        let index = rng.gen_range(0..values.len());
+        GTValue::arbitrary_from(rng, values[index])
+    }
+}
+
+impl ArbitraryFrom<Value> for GTValue {
+    fn arbitrary_from<R: Rng>(rng: &mut R, value: &Value) -> Self {
+        match value {
+            Value::Integer(i) => GTValue(Value::Integer(rng.gen_range(*i..i64::MAX))),
+            Value::Float(f) => GTValue(Value::Float(rng.gen_range(*f..1e10))),
+            Value::Text(t) => {
+                // Either lengthen the string, or make at least one character smaller and mutate the rest
+                let mut t = t.clone();
+                if rng.gen_bool(0.01) {
+                    t.push(rng.gen_range(0..=255) as u8 as char);
+                    GTValue(Value::Text(t))
+                } else {
+                    let mut t = t.chars().map(|c| c as u32).collect::<Vec<_>>();
+                    let index = rng.gen_range(0..t.len());
+                    t[index] += 1;
+                    // Mutate the rest of the string
+                    for i in (index + 1)..t.len() {
+                        t[i] = rng.gen_range('a' as u32..='z' as u32);
+                    }
+                    let t = t.into_iter().map(|c| char::from_u32(c).unwrap_or('a')).collect::<String>();
+                    GTValue(Value::Text(t))
+                }
+            }
+            Value::Blob(b) => {
+                // Either lengthen the blob, or make at least one byte smaller and mutate the rest
+                let mut b = b.clone();
+                if rng.gen_bool(0.01) {
+                    b.push(rng.gen_range(0..=255));
+                    GTValue(Value::Blob(b))
+                } else {
+                    let index = rng.gen_range(0..b.len());
+                    b[index] += 1;
+                    // Mutate the rest of the blob
+                    for i in (index + 1)..b.len() {
+                        b[i] = rng.gen_range(0..=255);
+                    }
+                    GTValue(Value::Blob(b))
+                }
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+enum Predicate {
+    And(Vec<Predicate>),        // p1 AND p2 AND p3... AND pn
+    Or(Vec<Predicate>),         // p1 OR p2 OR p3... OR pn
+    Eq(String, Value),          // column = Value
+    Gt(String, Value),          // column > Value
+    Lt(String, Value),          // column < Value
+}
+
+// This type represents the potential queries on the database.
+enum Query {
+    Create(Create),
+    Select(Select),
+    Insert(Insert),
+    Delete(Delete),
+}
+
+struct Create {
+    table: Table,
+}
+
+impl Arbitrary for Create {
+    fn arbitrary<R: Rng>(rng: &mut R) -> Self {
+        Create {
+            table: Table::arbitrary(rng),
+        }
+    }
+}
+
+struct Select {
+    table: String,
+    predicate: Predicate,
+}
+
+impl ArbitraryFrom<Vec<Table>> for Select {
+    fn arbitrary_from<R: Rng>(rng: &mut R, tables: &Vec<Table>) -> Self {
```

Comment:
```suggestion
    fn gen<R: Rng>(rng: &mut R, tables: &Vec<Table>) -> Self {
```
Nit? What do you think